### PR TITLE
Add download stop API for download service

### LINKED ISSUE #1242: feat: Model Pull has clear API and CLI to support Huggingface Repos
## Goal 

- `cortex model pull` should have clear APIs that support different model repo sources
- e.g. Huggingface, Cortex Hub

## Tasklist

- [x] #1393 
- [x] #1394 
- [x] #1395
- [x] #1398

## CLI
```
# Pulls immediately
cortex model pull <huggingface_url_sppecific_gguf>

# Lets user select quantization using CLI
cortex model pull <huggingface_url> 

# NOT SURE: Do we need an "info" equivalent?
# Gets repo type (e.g. GGUF, in future ONNX, TensorRT-LLM, dumps possible versions)
# Will power the "select quantization"
cortex model info <huggingface_url>
cortex model info <cortex_repo_url>    # Dumps tags
```

## API
- Is there any change necessary to the models pull API?
- https://cortex.so/api-reference/#tag/models/post/v1/models/%7BmodelId%7D/pull
```
- How do we represent Huggingface strings?
- How do we handle Huggingface Repos (i.e. where user has to select quantization)? 
```

## Key Questions

- Does Cortex need an API to get Huggingface Repo metadata?
- Does this need to be bubbled to Jan, to support the Huggingface Repo page?

## Linked Issues

- https://github.com/janhq/cortex.cpp/issues/1006
- https://github.com/janhq/cortex.cpp/issues/1145
- https://github.com/janhq/cortex.cpp/issues/1017
- And your PR: https://github.com/janhq/cortex.cpp/pull/1239

## Jan's Requirements

1. User enters Huggingface URL in import box
2. User clicks deep link from Huggingface

Cortex should support an API, that can support the following UI: 

Image

### LINKED ISSUE #1398: idea: [DownloadEvent] Adding download stop API
### Problem Statement

Currently we have no API for stop downloading

### Feature Idea

- Adding download stop API

### PR DESCRIPTION
**Pull request by namchuai**
## Describe Your Changes

- Refactor to add download stop API.
- Using multi interface from libcurl.
- Throttle event by 1 sec
- Return `DownloadTask` when calling download API.

## API:
```
curl --location --request DELETE 'http://127.0.0.1:39281/models/pull' \
--header 'Content-Type: application/json' \
--data '{
    "taskId": <your_download_task_id>
}'
```
Response when calling Download API
```
{
    "message": "Model start downloading!",
    "task": {
        "id": "Mistral-7B-Instruct-v0.1-GGUF",
        "items": [
            {
                "bytes": 3822024352,
                "checksum": "N/A",
                "downloadUrl": "https://huggingface.co/TheBloke/Mistral-7B-Instruct-v0.1-GGUF/resolve/main/mistral-7b-instruct-v0.1.Q3_K_L.gguf",
                "downloadedBytes": 0,
                "id": "TheBloke:Mistral-7B-Instruct-v0.1-GGUF:mistral-7b-instruct-v0.1.Q3_K_L.gguf",
                "localPath": "/Users/jamesnguyen/cortexcpp/models/huggingface.co/TheBloke/Mistral-7B-Instruct-v0.1-GGUF/mistral-7b-instruct-v0.1.Q3_K_L.gguf"
            }
        ],
        "type": "Model"
    }
}
```

## Issue
- Fix #1398 
- Fixes #1242

## Self Checklist

- [ ] Added relevant comments, esp in complex areas
- [ ] Updated docs (for bug fixes / features)
- [ ] Created issues for follow-up changes or refactoring needed

### CODE REVIEW

**Comment by dan-menlo** [comment]
@namchuai At the conclusion of this issue, please make sure we add this to the API Reference (i.e. swagger). cc @gabrielle-ong

File: engine/services/download_service.cc [hunk 1]
```diff
@@ -276,16 +218,196 @@ curl_off_t DownloadService::GetLocalFileSize(
...
+    DownloadTask task;
+    {
+      std::unique_lock<std::mutex> lock(queue_mutex_);
+      queue_condition_.wait(
+          lock, [this] { return !task_queue_.empty() || stop_flag_; });
+      if (stop_flag_) {
+        CTL_INF("Stopping download service..");
+        break;
+      }
+      if (!task_queue_.empty()) {
```
Comment (vansangpfiev) [inline engine/services/download_service.cc]: Not sure why we need to check empty() here?

Comment (namchuai) [reply engine/services/download_service.cc]: Good catch! It's unnecessary after I clean up that function. Removed.

File: engine/services/download_service.h [hunk 2]
```diff
@@ -21,19 +27,41 @@ class DownloadService {
...
+  ~DownloadService() {
+    if (event_queue_ != nullptr) {
+      stop_flag_ = true;
+      queue_condition_.notify_one();
+
+      CTL_INF("DownloadService is being destroyed.");
+      curl_multi_cleanup(multi_handle_);
+      curl_global_cleanup();
+
+      worker_thread_.join();
```
Comment (vansangpfiev) [inline engine/services/download_service.h]: I think should check if `worker_thread_` is joinable

File: engine/services/download_service.h [hunk 3]
```diff
@@ -42,57 +70,99 @@ class DownloadService {
...
 
-    auto active_task_id = service->active_download_task_id_;
-    auto active_item_id = service->active_download_item_id_;
-    if (!active_task_id.has_value() || !active_item_id.has_value()) {
-      return 0;
-    }
+  std::shared_ptr<DownloadingData> downloading_data_;
+
+  // active task that being downloaded atm
+  std::unordered_map<std::string, DownloadTask> active_tasks_;
```
Comment (vansangpfiev) [inline engine/services/download_service.h]: Seems like `active_tasks_` is unused?

**Review by vansangpfiev — APPROVED** [review]
Great work!

**Comment by gabrielle-ong** [comment]
@namchuai can I check if this new endpoint DELETE `models/pull` is added to Swagger?

**Pull request merged**